substrate/frame/metadata-hash-extension/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+//! compile time, it requires a little bit more setup. To have this environment variable available
+//! at compile time required to tell the `substrate-wasm-builder` to do so:
+#![doc = docify::embed!("src/tests.rs", enable_metadata_hash_in_wasm_builder)]
+//! As generating the metadata hash requires to compile the runtime twice, it is
+//! recommended to only enable the metadata hash generation when doing a build for a release or when
+//! you want to test this feature.
+
+extern crate alloc;
+/// For our tests
+extern crate self as frame_metadata_hash_extension;
+
+use alloc::vec::Vec;
+use codec::{Decode, Encode};
+use frame_support::DebugNoBound;
+use frame_system::Config;
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{DispatchInfoOf, SignedExtension},
+	transaction_validity::{TransactionValidityError, UnknownTransaction},
+};
+
+#[cfg(test)]
+mod tests;
+
+/// Type that encodes `None` to an empty vec.
+#[derive(Debug, PartialEq)]
+pub struct EncodeNoneToEmpty(pub Option<[u8; 32]>);
+
+impl Encode for EncodeNoneToEmpty {
+	fn encode(&self) -> Vec<u8> {
+		match self.0 {
+			Some(hash) => hash.encode(),
+			None => Vec::new(),
+		}
+	}
+}
+
+impl TypeInfo for EncodeNoneToEmpty {
+	type Identity = <[u8; 32] as TypeInfo>::Identity;
+
+	fn type_info() -> scale_info::Type {
+		<[u8; 32]>::type_info()
+	}
+}
+
+/// Extension for optionally checking the metadata hash.
+///
+/// The metadata hash is cryptographical representation of the runtime metadata. This metadata hash
+/// is build as described in [RFC78](https://polkadot-fellows.github.io/RFCs/approved/0078-merkleized-metadata.html).
+/// This metadata hash should give users the confidence that what they build with an online wallet
+/// is the same they are signing with their offline wallet and then applying on chain. To ensure
+/// that the online wallet is not tricking the offline wallet into decoding and showing an incorrect
+/// extrinsic, the offline wallet will include the metadata hash into the additional signed data and
+/// the runtime will then do the same. If the metadata hash doesn't match, the signature
+/// verification will fail and thus, the transaction will be rejected. The RFC contains more details
+/// on how it works.
+///
+/// The extension adds one byte (the `mode`) to the size of the extrinsic. This one byte is
+/// controlling if the metadata hash should be added to the signed data or not. Mode `0` means that
+/// the metadata hash is not added and `1` means that it is added. Further values of `mode` are
+/// reserved for future changes.
+///
+/// The metadata hash is read from the environment variable `RUNTIME_METADATA_HASH`. This
+/// environment variable is for example set by the `substrate-wasm-builder` when the feature for
+/// generating the metadata hash is enabled. If the environment variable is not set and `mode = 1`
+/// is passed, the transaction is rejected with [`UnknownTransaction::CannotLookup`].
```

Comment:
And when a transaction sets `mode = 0` but the runtime knows the hash? Is it optional or will it also be rejected?  
I guess we need to wait for wallet support first.

Reply:
The point of `mode = 0` being that you can disable it. Not everyone will need this and I don't see any reason to force people to use this.